File: server/src/lib/ratelimit.ts
/**
 * Spend protection for the one endpoint that costs money.
 *
 * Cairn deploys to a public URL with no login, so `/api/ask` is reachable by
 * anyone who finds it — and every miss behind it is a paid vision call. Three
 * independent limits apply, and a request has to clear all of them:
 *
 *   1. per-IP burst   — stops one client hammering it (a runaway loop, a stuck
 *                       retry, someone leaning on the button)
 *   2. per-IP hourly  — stops one client grinding steadily all afternoon
 *   3. global daily   — the actual wallet guard. Caps total spend per day no
 *                       matter how many distinct clients show up, which is the
 *                       only limit that helps against a small crowd or a spread
 *                       of addresses.
 *
 * Only *model* calls are counted. Recall hits cost nothing, so they're never
 * limited — which means the team-memory layer isn't just a latency win, it
 * takes real pressure off this budget.
 *
 * Storage mirrors TrailStore: in-memory by default so the app runs anywhere,
 * Upstash when configured. Be honest about what the memory adapter is worth on
 * serverless — counters live per instance and reset on cold start, so the real
 * ceiling is somewhat above the configured one. It stops loops and casual
 * abuse; it is not an airtight budget. Set the Upstash variables for a limit
 * that actually holds across instances.
 */

export interface RateLimitConfig {
  /** Model calls allowed per IP per minute. */
  perIpPerMinute: number;
  /** Model calls allowed per IP per hour. */
  perIpPerHour: number;
  /** Model calls allowed across everyone per day. The wallet guard. */
  globalPerDay: number;
}

function envInt(name: string, fallback: number): number {
  const raw = process.env[name];
  if (!raw) return fallback;
  const n = Number.parseInt(raw, 10);
  return Number.isFinite(n) && n > 0 ? n : fallback;
}

/**
 * Defaults are sized against a demo budget, not against production traffic.
 * At roughly 3 cents a call, 100/day caps exposure near $3/day — enough for a
 * reviewer to explore thoroughly, low enough that a scraper can't drain the
 * account overnight.
 */
export function getConfig(): RateLimitConfig {
  return {
    perIpPerMinute: envInt("RATE_LIMIT_PER_IP_PER_MINUTE", 6),
    perIpPerHour: envInt("RATE_LIMIT_PER_IP_PER_HOUR", 30),
    globalPerDay: envInt("RATE_LIMIT_GLOBAL_PER_DAY", 100),
  };
}

export interface RateLimitResult {
  allowed: boolean;
  /** Human-facing explanation. Shown directly in the UI. */
  message?: string;
  /** Seconds until the caller should try again. Sent as Retry-After. */
  retryAfter?: number;
  /** Which limit tripped — useful in logs when tuning. */
  limit?: "burst" | "hourly" | "daily";
}

/* ------------------------------------------------------------------ stores */

interface CounterStore {
  /** Increment `key`, creating it with `windowSec` TTL if absent. Returns the new count. */
  bump(key: string, windowSec: number): Promise<number>;
}

class MemoryCounterStore implements CounterStore {
  private get counters(): Map<string, { count: number; expiresAt: number }> {
    const g = globalThis as typeof globalThis & {
      __cairnRate?: Map<string, { count: number; expiresAt: number }>;
    };
    if (!g.__cairnRate) g.__cairnRate = new Map();
    return g.__cairnRate;
  }

  async bump(key: string, windowSec: number): Promise<number> {
    const now = Date.now();
    const counters = this.counters;

    // Opportunistic sweep. Without it a long-lived instance accumulates one
    // entry per IP per window forever, which is a slow memory leak.
    if (counters.size > 5000) {
      for (const [k, v] of counters) if (v.expiresAt <= now) counters.delete(k);
    }

    const existing = counters.get(key);
    if (!existing || existing.expiresAt <= now) {
      counters.set(key, { count: 1, expiresAt: now + windowSec * 1000 });
      return 1;
    }
    existing.count += 1;
    return existing.count;
  }
}

class UpstashCounterStore implements CounterStore {
  constructor(
    private url: string,
    private token: string,
  ) {}

  async bump(key: string, windowSec: number): Promise<number> {
    // Pipelined so INCR and EXPIRE cost one round trip rather than two.
    // EXPIRE is set every time rather than only on creation: it's idempotent
    // enough here (a fixed window that slides slightly is fine) and it avoids
    // a key that somehow lost its TTL blocking a caller forever.
    const res = await fetch(`${this.url}/pipeline`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${this.token}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify([
        ["INCR", key],
        ["EXPIRE", key, String(windowSec), "NX"],
      ]),
      cache: "no-store",
    });
    if (!res.ok) throw new Error(`Upstash ${res.status}`);
    const body = (await res.json()) as Array<{ result: number }>;
    return body[0]?.result ?? 1;
  }
}

let cachedStore: CounterStore | null = null;

function getCounterStore(): CounterStore {
  if (cachedStore) return cachedStore;
  const url = process.env.UPSTASH_REDIS_REST_URL;
  const token = process.env.UPSTASH_REDIS_REST_TOKEN;
  cachedStore = url && token ? new UpstashCounterStore(url, token) : new MemoryCounterStore();
  return cachedStore;
}

/** True when limits are shared across instances rather than per-instance. */
export function isRateLimitDurable(): boolean {
  return Boolean(process.env.UPSTASH_REDIS_REST_URL && process.env.UPSTASH_REDIS_REST_TOKEN);
}

/* ------------------------------------------------------------------- check */

/** Day bucket in UTC, so the global window is deterministic across regions. */
function dayKey(): string {
  return new Date().toISOString().slice(0, 10);
}

/**
 * Consumes one unit against every limit. Call this only when a model call is
 * actually about to happen — never on a recall hit, which is free.
 *
 * Counters are incremented even when a later limit rejects the request. That's
 * deliberate: a client that keeps retrying through a 429 should stay blocked
 * rather than being handed a fresh allowance, and the alternative (rolling back
 * the increments) is racy for no real benefit.
 */
export async function consumeModelCall(ip: string): Promise<RateLimitResult> {
  const cfg = getConfig();
  const store = getCounterStore();

  try {
    const [minute, hour, day] = await Promise.all([
      store.bump(`cairn:rl:m:${ip}:${Math.floor(Date.now() / 60_000)}`, 60),
      store.bump(`cairn:rl:h:${ip}:${Math.floor(Date.now() / 3_600_000)}`, 3600),
      store.bump(`cairn:rl:d:${dayKey()}`, 86_400),
    ]);

    if (minute > cfg.perIpPerMinute) {
      return {
        allowed: false,
        limit: "burst",
        retryAfter: 60,
        message: "Slow down a moment — that's a lot of questions at once. Try again in a minute.",
      };
    }
    if (hour > cfg.perIpPerHour) {
      return {
        allowed: false,
        limit: "hourly",
        retryAfter: 900,
        message:
          "You've hit the hourly limit for live answers. Saved trails still work — try the Trails tab.",
      };
    }
    if (day > cfg.globalPerDay) {
      return {
        allowed: false,
        limit: "daily",
        retryAfter: 3600,
        message:
          "Cairn has reached its daily budget for live answers. Browsing and replaying trails still works.",
      };
    }

    return { allowed: true };
  } catch {
    // A limiter outage must not take the product down, and must not silently
    // remove the spend cap either. Failing open is the lesser evil here: the
    // per-request cost is bounded and small, whereas failing closed would make
    // an Upstash blip look like a broken product to a reviewer.
    return { allowed: true };
  }
}

/**
 * Best-effort client address. Vercel populates x-forwarded-for; the first entry
 * is the original client. Falls back to a shared bucket, which means unknown
 * clients throttle each other — acceptable, since the global daily cap is the
 * limit that actually protects the budget.
 */
export function clientIp(req: Request): string {
  const fwd = req.headers.get("x-forwarded-for");
  if (fwd) return fwd.split(",")[0]!.trim();
  return req.headers.get("x-real-ip")?.trim() || "unknown";
}

File: internal/cairn/ratelimit.go
package cairn

import (
	"net"
	"net/http"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

// Spend protection for the endpoints that cost money.
//
// The server is public and unauthenticated, so /api/ask and /api/transcribe are
// reachable by anyone who finds the URL, and both spend real money per call.
// Three limits apply and a request must clear all of them:
//
//	per-IP burst   stops one client hammering it — a runaway loop, a stuck retry
//	per-IP hourly  stops one client grinding steadily all afternoon
//	global daily   the wallet guard, and the only one that helps against a crowd
//
// Only calls that actually reach a paid provider are counted. Recall hits are
// free and never consume quota, which means the team-memory layer isn't just a
// latency win — it takes real pressure off this budget.
//
// Unlike the serverless version this replaces, these counters live in one
// process, so the limits genuinely hold rather than resetting per instance.

type rateLimits struct {
	perIPPerMinute int
	perIPPerHour   int
	globalPerDay   int
}

func loadRateLimits() rateLimits {
	// Sized against a demo budget rather than production traffic: at roughly
	// three cents a vision call, 100/day caps exposure near $3/day — enough for
	// a reviewer to explore properly, low enough that a scraper can't drain the
	// account overnight.
	return rateLimits{
		perIPPerMinute: envInt("RATE_LIMIT_PER_IP_PER_MINUTE", 6),
		perIPPerHour:   envInt("RATE_LIMIT_PER_IP_PER_HOUR", 30),
		globalPerDay:   envInt("RATE_LIMIT_GLOBAL_PER_DAY", 100),
	}
}

func envInt(name string, fallback int) int {
	if raw := os.Getenv(name); raw != "" {
		if n, err := strconv.Atoi(raw); err == nil && n > 0 {
			return n
		}
	}
	return fallback
}

type counter struct {
	count     int
	expiresAt time.Time
}

type rateLimiter struct {
	mu       sync.Mutex
	counters map[string]*counter
	limits   rateLimits
}

func newRateLimiter() *rateLimiter {
	rl := &rateLimiter{
		counters: make(map[string]*counter),
		limits:   loadRateLimits(),
	}
	go rl.sweep()
	return rl
}

// sweep drops expired counters. Without it, one entry per IP per window
// accumulates forever — a slow leak in a long-running process.
func (rl *rateLimiter) sweep() {
	for range time.Tick(5 * time.Minute) {
		now := time.Now()
		rl.mu.Lock()
		for k, c := range rl.counters {
			if now.After(c.expiresAt) {
				delete(rl.counters, k)
			}
		}
		rl.mu.Unlock()
	}
}

func (rl *rateLimiter) bump(key string, window time.Duration) int {
	now := time.Now()
	c, ok := rl.counters[key]
	if !ok || now.After(c.expiresAt) {
		rl.counters[key] = &counter{count: 1, expiresAt: now.Add(window)}
		return 1
	}
	c.count++
	return c.count
}

type limitVerdict struct {
	Allowed    bool
	Message    string
	RetryAfter int
}

// consume charges one unit against every limit. Call this only when a paid call
// is actually about to happen — never on a recall hit.
//
// Counters increment even when a later limit rejects the request. That's
// deliberate: a client retrying through a 429 should stay blocked rather than
// being handed a fresh allowance, and rolling back would be racy for no gain.
func (rl *rateLimiter) consume(ip string) limitVerdict {
	now := time.Now()
	rl.mu.Lock()
	defer rl.mu.Unlock()

	minute := rl.bump("m:"+ip+":"+strconv.FormatInt(now.Unix()/60, 10), time.Minute)
	hour := rl.bump("h:"+ip+":"+strconv.FormatInt(now.Unix()/3600, 10), time.Hour)
	day := rl.bump("d:"+now.UTC().Format("2006-01-02"), 24*time.Hour)

	switch {
	case minute > rl.limits.perIPPerMinute:
		return limitVerdict{Message: "Slow down a moment — that's a lot of questions at once. Try again in a minute.", RetryAfter: 60}
	case hour > rl.limits.perIPPerHour:
		return limitVerdict{Message: "You've hit the hourly limit for live answers. Saved trails still work.", RetryAfter: 900}
	case day > rl.limits.globalPerDay:
		return limitVerdict{Message: "Cairn has reached its daily budget for live answers. Browsing and replaying trails still works.", RetryAfter: 3600}
	}
	return limitVerdict{Allowed: true}
}

// clientIP is best-effort. Proxies populate X-Forwarded-For and the first entry
// is the original client. Falls back to a shared bucket, which means unknown
// clients throttle each other — acceptable, since the global daily cap is what
// actually protects the budget.
func clientIP(r *http.Request) string {
	if fwd := r.Header.Get("X-Forwarded-For"); fwd != "" {
		return strings.TrimSpace(strings.Split(fwd, ",")[0])
	}
	if real := r.Header.Get("X-Real-IP"); real != "" {
		return strings.TrimSpace(real)
	}
	if host, _, err := net.SplitHostPort(r.RemoteAddr); err == nil {
		return host
	}
	return "unknown"
}
